src/util/metadata/side_metadata/global.rs

```diff
@@ -946,6 +1075,194 @@ impl SideMetadataSpec {
+            // This won't be used for VO bit, as VO bit is global and is always contiguous. So for now, I am not bothered to do it.
+            warn!("We are trying to search non zero bits in an discontiguous side metadata. The performance is slow, as MMTk does not optimize for this case.");
+            self.find_prev_non_zero_value_simple::<T>(data_addr, search_limit_bytes)
+        }
+    }
+
+    fn find_prev_non_zero_value_simple<T: MetadataValue>(
+        &self,
+        data_addr: Address,
+        search_limit_bytes: usize,
+    ) -> Option<Address> {
+        let region_bytes = 1 << self.log_bytes_in_region;
+        // Figure out the range that we need to search.
+        let start_addr = data_addr.align_down(region_bytes);
+        let end_addr = data_addr
+            .saturating_sub(search_limit_bytes)
+            .align_down(region_bytes);
+
+        let mut cursor = start_addr;
+        while cursor > end_addr {
+            // We encounter an unmapped address. Just return None.
+            if !cursor.is_mapped() {
+                return None;
+            }
+            // If we find non-zero value, just return it.
+            if !unsafe { self.load::<T>(cursor).is_zero() } {
+                return Some(cursor);
+            }
+            cursor -= region_bytes;
+        }
+        None
+    }
+
+    #[allow(clippy::let_and_return)]
+    fn find_prev_non_zero_value_fast<T: MetadataValue>(
+        &self,
+        data_addr: Address,
+        search_limit_bytes: usize,
+    ) -> Option<Address> {
+        debug_assert!(self.uses_contiguous_side_metadata());
+
+        // Quick check if the data address is mapped at all.
+        if !data_addr.is_mapped() {
+            return None;
+        }
+        // Quick check if the current data_addr has a non zero value.
+        if !unsafe { self.load::<T>(data_addr).is_zero() } {
+            return Some(data_addr.align_down(1 << self.log_bytes_in_region));
+        }
+
+        // Figure out the start and end data address.
+        let start_addr = data_addr.saturating_sub(search_limit_bytes) + 1usize;
+        let end_addr = data_addr;
+
+        // Then figure out the start and end metadata address and bits.
+        let start_meta_addr = address_to_contiguous_meta_address(self, start_addr);
+        let start_meta_shift = meta_byte_lshift(self, start_addr);
+        let end_meta_addr = address_to_contiguous_meta_address(self, end_addr);
+        let end_meta_shift = meta_byte_lshift(self, end_addr);
+
+        // The result will be set by one of the following closures.
+        // Use Cell so it doesn't need to be mutably borrowed by the two closures which Rust will complain.
+        let res = std::cell::Cell::new(None);
+
+        // Check and find the last bit that is set. We try load words where possible, and fall back to load bytes.
+        let check_bytes_backwards = |start: Address, end: Address| -> bool {
```

Comment:
This is quite a big lambda.  I think it should be a very common operation to search for the last non-zero byte in an array, and I believe there are existing third-party crates that can do this, and probably with SIMD support.

If we want to implement it by ourselves, we may extract this into some utility modules, and implement this algorithm over anything that supports "loading at word or byte granularity".  That should include the side metadata, a `Vec<u8>`, or even a `String`, so that it will be easy to write unit tests for this algorithm without depending on the side metadata framework.

By the way, if we want to be aggressive and fully commit into search or updating side metadata bitmaps word by word, we can replace `check_bytes_backwards` and `check_bits_backwards` with `check_words_backwards` and `check_bits_in_word_backwards`.  But that itself may deserve a separate PR.

Reply:
I thought about it. It requires changing `iterate_meta_bits`, and all the methods that use them. But most methods don't need to 'check words'. We could also add a similar function that allows 'check words', but it will duplicate code or have a very similar pattern as `iterate_meta_bits`. Neither sounds ideal. So I think it is probably a better idea to implement 'checking word' with 'check bytes'. 

I could extract some code from the closure and make it reusable.

Reply:
I extract those code to be separate functions in `side_metadata::helpers`. They are not general to be used for things like `Vec<u8>`, as we need to check if the metadata address is mapped or not which only applies to our side metadata.

Reply:
> ... They are not general to be used for things like `Vec<u8>`, as we need to check if the metadata address is mapped or not which only applies to our side metadata.

This is easy to solve.  We require the bit holder to implement a trait that must provide a `is_mapped` function, like this:

```rust
trait HoledByteArray {
    fn load_byte(&self, address: Address) -> u8;
    fn load_word(&self, address: Address) -> usize;
    fn is_mapped(&self, address: Address) -> bool;
}
```

And we can implement it for `Vec<u8>`, using byte index as `address`, and make `is_mapped` always return true.

```rust
impl HoledByteArray for Vec<u8> {
    fn load_byte(&self, address: Address) -> u8 { self[address.as_usize()] }
    fn load_word(&self, address: Address) -> usize { /* load 8 consecutive elements */ }
    fn is_mapped(&self, address: Address) -> bool { true }
```

And we can implement it for metadata, too.  We just load from raw address and actually check the mapping.

Reply:
Sounds good. But maybe we don't need to add it now: it adds complexity and there is no clear sign that we will use it any time soon. When anyone needs to reuse the code, they can do the refactoring and make the method more general.